input-profile.ts

```diff
@@ -0,0 +1,311 @@
+                .filter((c: any) => !this.toFilter.has(c.type));
+            components.push(...comps);
+            const children = currentObj.children || [];
+            // Push children onto the stack in reverse order to maintain the correct order
+            for (let i = children.length - 1; i >= 0; --i) {
+                stack.push(children[i]);
+            }
+        }
+
+        return components;
+    }
+
+    setComponentsActive(active: boolean) {
+        const comps = this.defaultControllerComponents;
+        if (!comps) return;
+        for (let i = 0; i < comps.length; ++i) {
+            comps[i].active = active;
+        }
+    }
+
+    onInputSourcesChange(event: any) {
+        if (this.modelLoaded && !this.mapToDefaultController) {
+            this.setComponentsActive(false);
+        }
+
+        event.added.forEach((xrInputSource: any) => {
+            if (xrInputSource.hand != null) return;
+            if (this.handedness === xrInputSource.handedness) {
+                var profile = this.path + xrInputSource.profiles[0];
+                if (this.customProfileFolder !== '') {
+                    profile = this.customProfileFolder;
+                }
+
+                this.url = profile + '/profile.json';
+                console.log('url is ' + this.url);
+
+                fetch(this.url)
+                    .then((res) => res.json())
+                    .then((out) => {
+                        this.ProfileJSON = out;
+                        console.log(this.ProfileJSON);
+                        const consoleText = this.modelLoaded
+                            ? 'Profile loaded' + this.url
+                            : 'Reloaded Profile for gamepad mapping ';
+                        console.log(consoleText);
+                        this.loadGamepad(profile, xrInputSource);
+                    })
+                    .catch((e) => {
+                        console.error('failed to load profile.json');
+                        console.error(e);
+                    });
+            }
+        });
+    }
+
+    isModelLoaded() {
+        return this.controllerModel !== null;
+    }
+
+    loadGamepad(profile: string, xrInputSource: any) {
+        this.gamepad = xrInputSource.gamepad;
+
+        const assetPath = profile + '/' + this.handedness + '.glb';
+        if (this.modelLoaded) return;
+        console.log('flag value == ');
+        console.log(this.mapToDefaultController);
+
+        if (!this.mapToDefaultController) {
+            this.engine.scene
+                .append(assetPath)
+                .then((obj: any) => {
+                    this.controllerModel = obj;
+                    this.setComponentsActive(false);
+                    console.log('setting comp false ');
+                    this.controllerModel.parent = this.object;
+                    this.controllerModel.setTranslationLocal([0, 0, 0]);
+                    this.getGamepadObjectsFromProfile(
```

Comment:
The code here and the code starting line `203` are identitcal, you should find a way to basically first dowenload the model if `!this.mapToDefaultController`, or just go through the code setting up the profile. No need to duplicate that part